### PR TITLE
PARQUET-2436: More optimal memory usage in compression codecs

### PR DESCRIPTION
**Pull request by gszadovszky**
Rewritten direct codec implementations to have common memory handling

Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET-2436) issues and references
  them in the PR title. For example, "PARQUET-1234: My Parquet PR"
    - https://issues.apache.org/jira/browse/PARQUET-XXX
    - In case you are adding a dependency, check if the license complies with
      the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines
  from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    1. Subject is limited to 50 characters (not including Jira issue reference)
    1. Subject does not end with a period
    1. Subject uses the imperative mood ("add", not "adding")
    1. Body wraps at 72 characters
    1. Body explains "what" and "why", not "how"

### Style
- [x] My contribution adheres to the code style guidelines and Spotless passes.
    - To apply the necessary changes, run `mvn spotless:apply -Pvector-plugins`

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecFactory.java [hunk 1]
```diff
@@ -250,75 +341,86 @@ public BytesInput decompress(BytesInput bytes, int uncompressedSize) throws IOEx
...
 
-      outgoing.limit(size);
+    @Override
+    void closeDecompressor() {
+      context.close();
+    }
+  }
+
+  private class ZstdCompressor extends BaseCompressor {
+    private final ZstdCompressCtx context;
```
Comment (wgtmac) [inline parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecFactory.java:398]: I'm not familiar with `ZstdCompressCtx`. Does it provide the same performance as before? IIUC, original impl leverages zstd workers but I didn't see it from the new class.

https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectZstd.java#L91

Comment (gszadovszky) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecFactory.java:398]: There is a big difference between our "usual" codecs that are using the Hadoop codecs and the ones created by the DirectCodecFactory. The first one has a stream interface behind our interfaces.
At the first attempt of implementing the direct Zstd codec, I was using the stream implementation for the BytesInput and a static method of the Zstd API for the ByteBuffer interface.
Now, I realized that when we use a direct ByteBufferAllocator the BytesInput objects are also mainly represented by direct ByteBuffers. So we can use them without a stream.
The context is used by the static API of the Zstd codec anyway. Based on the class comments it is suggested to use the context instead.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecFactory.java [hunk 2]
```diff
@@ -186,6 +174,100 @@ public void release() {
     }
   }
 
+  private abstract class BaseDecompressor extends BytesDecompressor {
```
[the remark this answers is not in the document]
Comment (gszadovszky) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecFactory.java:177]: Unfortunately, the CodecFactory's interface still produces BytesInputDecompress/Compressor so I cannot change it here without rewriting the API causing breaking changes.

**Review by wgtmac — APPROVED** [review]
LGTM

**Comment by gszadovszky** [comment]
Thanks for the review, @wgtmac!

**Pull request merged**